### ISSUE TITLE
the formula for layout optimization is slightly different from that in UMAP paper?

### ISSUE THREAD

**Issue by overshiki** [lmcinnes/umap#460]
Hi,
Just trying to understand the layout optimization part of code, I find that the formula for positive edges here:
`                if dist_squared > 0.0:
                    grad_coeff = -2.0 * a * b * pow(dist_squared, b - 1.0)
                    grad_coeff /= a * pow(dist_squared, b) + 1.0
                else:
                    grad_coeff = 0.0`
is different from what explained in https://arxiv.org/pdf/1802.03426.pdf chapter 3.2:
![image](https://user-images.githubusercontent.com/8148439/86222358-b4ce0a80-bbb8-11ea-94f0-33f58b34d25d.png)
I understand the W is missing because we are doing importance sampling, but the denominator is also slightly different.

Also for negative edges, we do something like this:
`                    if dist_squared > 0.0:
                        grad_coeff = 2.0 * gamma * b
                        grad_coeff /= (0.001 + dist_squared) * (
                            a * pow(dist_squared, b) + 1
                        )
                    elif j == k:
                        continue
                    else:
                        grad_coeff = 0.0`

but in the paper, it is:
![image](https://user-images.githubusercontent.com/8148439/86222735-3756ca00-bbb9-11ea-9fab-a8c3d1d70041.png)

Am I missing something? Or perhaps the version in the code works better?

Thanks a lot in advance

**Comment by lmcinnes** [comment]
Hi -- there were some transcription errors in the paper, which have been corrected elsewhere but haven't been uploaded as a new version to ArXiv yet. Hopefully that will come soon.

**Comment by jlmelville** [comment]
This is my fault, I wrote this bit of the paper and I failed to generate the correct LaTeX for the equations. The code is correct and the paper is wrong. I am very sorry for the confusion.

**Comment by overshiki** [comment]
Thanks a lot for the clarification, that is helpful.
Also I observe that the negative(repulsive) sampling is inside the positive(attractive) sampling loop, so I assume the positive(attractive) sampling rate is w, and the negative(repulsive) sampling rate per positive loop should be (1-w)/w. Am I right about this?

**Comment by jlmelville** [comment]
@overshiki I don't entirely follow the argument about the sampling rate for negatives. I can state how it works conceptually (although the actual implementation touches on your second question):

1. You sample a positive edge with a probability proportional to weight of that edge (w), so you now have two vertices, p_1, p_2.
2. Then you sample any of the N vertices with a uniform probability (where N is the total number of observations in the dataset), to get a negative vertex n_1. The negative update uses the negative edges formed from (p_1, n_1) and (p_2, n_1), and the weights of these edges are zero.
3. Repeat step 2 until you have done it `negative_sample_rate` times.
4. Return to step 1.

Step 1 of the sampling would imply that there is some sort of probabilistic sampling of the positive edges, but as you've noted, this isn't exactly what happens. @lmcinnes can explain this better, but I'm pretty sure this is not really intended as an annealing technique, but just a result of using the expected number of times you'd sample the edges, relative to sampling the edge with the largest weight in every epoch.

For example, if you use 100 epochs, then the edge with the largest weight (for simplicity, let's say its weight is 1.0) gets sampled in all 100 epochs. For an edge with a weight of 0.5, it will get sampled in 50 of the epochs. An edge with a weight of 0.1 gets samples in 10 epochs and so on. I should also note that when an edge is included in an epoch, it doesn't always get exactly `negative_sample_rate` negative samples, but overall the total number of negative samples gives the expected `negative_sample_rate`.

The decreasing learning rate is just a common way to get stochastic gradient descent to converge.

Hope this helps.

**Comment by overshiki** [comment]
@jlmelville Sorry for the confusing question, and thank you for the detailed explanation! I am now confirmed that the edges are sampled with different frequencies according to their weights W. The reason why I thought it was kind of annealing technique is that I experimentally observed that the last few epochs of the training always give the most significant improvement over the learned embedding. And since edges are sampled with different frequencies, the first time each edge is involved in the training is different(some have longer time of delay), I initially guess these may be related to the above mentioned phenomenon.  I now see this is a misunderstanding.
It seems the negative edges are just randomly sampled and does not following the 1-w weights as described in the paper. I understand that since most of the edges are negative(our knn-based graph are sparse), 1-w will be 1 in most of the time. In this case, putting negative sampling loop outside of the positive sampling loop will also work fine and may provide better parallel properties.
Thanks a lot and you guys are excellent!

**Comment by lwiklendt** [comment]
There is another discrepency between the code and paper, but this time it looks like the code might be wrong. As of definition 11 in the paper, Phi is the differentiable approximation to Psi, and is defined as 1/(1 + a (||x-y||^2)^b). In Algorithm 5, the gradient of the log of Phi is used to update the embedding, and in the code the numerator for `grad_coeff` contains `pow(dist_squared, b - 1.0)`. However, if we pull out the square into the power then we would get something like `pow(np.sqrt(dist_squared), 2.0 * (b - 1.0))` = `pow(np.sqrt(dist_squared), 2.0 * b - 2.0)`, but the correct version should be `pow(np.sqrt(dist_squared), 2.0 * b - 1.0)`. See [gradient](https://www.sympygamma.com/input/?i=simplify%28diff%28log%281%2F%281+%2B+a*%28x-y%29**%282*b%29%29%29%2C+x%29%29) [calculation](https://www.wolframalpha.com/input/?i=grad%28log%281%2F%281+%2B+a+%28x-y%29%5E%282b%29%29%29%2C+x%29). To correct this, the numerator should really be `pow(dist_squared, b - 0.5)`.

**Comment by jlmelville** [comment]
I see that a new version of the UMAP (v3) paper is on arxiv, which corrects the worst mistakes I made (or at least that have been found so far) and which were spotted by @overshiki in this issue. Phew.

@lwiklendt I'm a bit confused by the gradient calculations you link to. You use `x-y`, where I would expect to see the Euclidean distance between those two vectors.

FWIW I have calculated the UMAP gradient (and related methods) by hand on multiple occasions, and I am highly confident what's in the code is correct. Because it is quite confusing and I regularly forget the details, I have documented it in a bit of detail in a few places (e.g. [here](https://jlmelville.github.io/uwot/umap-for-tsne.html) and [here](https://jlmelville.github.io/smallvis/theory.html)).

**Comment by lwiklendt** [comment]
@jlmelville yes, of course, my mistake.

    from sympy import *
    var('x1 x2 y1 y2 a b dist_squared', real=True)
    x = Matrix([x1, x2])
    y = Matrix([y1, y2])
    d = (x-y).norm(2)
    dlogphi_dx = log(1/(1 + a * d**(2*b))).diff(x)
    dlogphi_dx = dlogphi_dx.subs((x-y).norm(2)**2, dist_squared)
    pprint(simplify(dlogphi_dx), use_unicode=False)
